bundles/org.openhab.binding.mielecloud/src/main/java/org/openhab/binding/mielecloud/internal/auth/OpenHabOAuthTokenRefresher.java

```diff
@@ -0,0 +1,153 @@
+    }
+
+    @Override
+    public void unsetRefreshListener(String serviceHandle) {
+        final AccessTokenRefreshListener refreshListener = listenerByServiceHandle.get(serviceHandle);
+        if (refreshListener != null) {
+            try {
+                OAuthClientService clientService = getOAuthClientService(serviceHandle);
+                clientService.removeAccessTokenRefreshListener(refreshListener);
+            } catch (OAuthException e) {
+                logger.warn("Failed to remove refresh listener: OAuth client service is unavailable.");
+                logger.debug("Exception details:", e);
+            }
+        }
+        listenerByServiceHandle.remove(serviceHandle);
+    }
+
+    @Override
+    public void refreshToken(String serviceHandle) {
+        if (listenerByServiceHandle.get(serviceHandle) == null) {
+            logger.warn("Token refreshing was requested but there is no token refresh listener registered!");
+            return;
+        }
+
+        OAuthClientService clientService = getOAuthClientService(serviceHandle);
+        refreshAccessToken(clientService);
+    }
+
+    private OAuthClientService getOAuthClientService(String serviceHandle) {
+        final OAuthFactory oauthFactory = this.oauthFactory;
+        if (oauthFactory == null) {
+            throw new OAuthException("OAuth factory is not available.");
+        }
+
+        final OAuthClientService clientService = oauthFactory.getOAuthClientService(serviceHandle);
+        if (clientService == null) {
+            throw new OAuthException("OAuth client service is not available.");
+        }
+        return clientService;
+    }
+
+    private void refreshAccessToken(OAuthClientService clientService) {
+        try {
+            final AccessTokenResponse accessTokenResponse = clientService.refreshToken();
+            final String accessToken = accessTokenResponse.getAccessToken();
+            if (accessToken == null) {
+                throw new OAuthException("Access token is not available.");
+            }
+        } catch (org.openhab.core.auth.client.oauth2.OAuthException e) {
+            throw new OAuthException("An error occured during token refresh: " + e.getMessage(), e);
+        } catch (IOException e) {
+            throw new OAuthException("A network error occured during token refresh: " + e.getMessage(), e);
+        } catch (OAuthResponseException e) {
+            throw new OAuthException("Miele cloud service returned an illegal response: " + e.getMessage(), e);
+        }
+    }
+
+    @Override
+    public Optional<String> getAccessTokenFromStorage(String serviceHandle) {
+        try {
+            AccessTokenResponse tokenResponse = getOAuthClientService(serviceHandle).getAccessTokenResponse();
+            if (tokenResponse == null) {
+                return Optional.empty();
+            } else {
+                return Optional.of(tokenResponse.getAccessToken());
+            }
+        } catch (Exception e) {
```

Comment:
We expect multiple exceptions here:
- `org.openhab.binding.mielecloud.internal.auth.OAuthException`: If the `OAuthClientService` is not available
- `IOException`: If a network/IO issue occurs while getting the access token response
- `OAuthErrorException` / `OAuthResponseException`: If an OAUTH error response was returned by the cloud while getting the access token response
- `OAuthException`: For other exceptions that occurred  while getting the access token response


The latter three come from `OAuthClientService.getAccessTokenResponse()` (btw. the documentation seems to be outdated there). We want to handle all these exceptions the same way and their common parent is, well, `Exception`.

Reply:
In general you want to avoid catching Exception unless you are handling all cases of Exception.  If you have multiple exception types and want to handle them the same way you use multi-catch:
`catch (OAuthException | IOException | OAuthErrorException | OAuthResponseException( { `

Reply:
I agree to @boc-tothefuture . Catching `Exception` catches also all `RuntimeException`s